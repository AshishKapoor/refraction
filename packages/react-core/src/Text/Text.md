The `Text` component is a little utility that makes it easy
to style any text with any of the supported text styles provided
by [`@quid/theme#textStyles`](#theme).

```
<Text type="large bold">Large bold text here</Text>
```

```
// The bold style is inherited by the default H1 styles of this web page
<Text as="h1" type="normal">Normal text with H1 tag</Text>
```

This component provides a complementary skeleton, it can be used as placeholder
while the actual content is still loading, the following examples are rendering
the same content of the two above, but with the skeleton component:

```
<Text.Skeleton type="large bold">Large bold text here</Text.Skeleton>
```

```
<Text.Skeleton as="h1" type="normal">Normal text with H1 tag</Text.Skeleton>
```

Alternatively, you can provide a fixed width to an empty skeleton:

```
<Text.Skeleton type="normal" width="15em" />
```

#### Migration from @quid/react-components#Text

If you were a user of the previous component, there are a few breaking changes:

1. Text.H1, H2, P, etc... are not supported anymore, use the `as` property instead;
2. The skeleton component now takes the same amount of space of the Text base component;
3. The skeleton component will render sensibly differently, take extra attention during migration;
