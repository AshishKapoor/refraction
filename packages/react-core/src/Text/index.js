/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import * as React from 'react';
import { textStyles, withFallback as wf } from '@quid/theme';
import { keyframes } from '@emotion/core';
import css from '@emotion/css/macro';
import styled from '@emotion/styled/macro';
import Color from 'color';

type Props = {
  /** Any valid HTML tag name used to render the text block */
  as?: string,
  /** A space separated list of supported text styles */
  type?: string,
};

type SkeletonProps = {
  ...Props,
  width?: string,
};

const BACKGROUND = {
  light: 'gray2',
  dark: 'gray7',
};

const reduceOpacity = color =>
  Color(color)
    .alpha(0.5)
    .string();

const pulse = wf(
  props => keyframes`
  0% {
    background-color: ${props.theme.colors[BACKGROUND[props.theme.current]]};
  }

  50% {
    background-color: ${reduceOpacity(
      props.theme.colors[BACKGROUND[props.theme.current]]
    )};
  }

  100% {
    background-color: ${props.theme.colors[BACKGROUND[props.theme.current]]};
  }
`
);

// @component
const Skeleton: React.StatelessFunctionalComponent<SkeletonProps> = styled(
  ({ as: As = 'span', type, className, width, ...props }) => {
    return (
      <As {...props}>
        <span className={className}>{props.children}</span>
      </As>
    );
  }
)`
  ${({ type = '' }) => textStyles(...type.split(' '))};
  color: transparent;
  user-select: none;
  border-radius: 2px;
  background-color: ${wf(
    props => props.theme.colors[BACKGROUND[props.theme.current]]
  )};
  animation: ${pulse} 3s ease-in-out infinite;
  ${wf(
    props =>
      props.width != null &&
      css`
        width: ${props.width};
        display: inline-block;
        &::before {
          content: 'X';
        }
      `
  )}
`;

// @component
const Text: React.StatelessFunctionalComponent<Props> & {
  Skeleton: typeof Skeleton,
} = styled.span`
  ${({ type = '' }) => textStyles(...type.split(' '))}
`;

Text.Skeleton = Skeleton;

export { Skeleton };

// @component
export default Text;
