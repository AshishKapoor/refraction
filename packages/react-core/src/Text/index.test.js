/**
 * Copyright (c) Quid, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
// @flow
import React from 'react';
import { mount } from 'enzyme';
import Text from './';

it('renders the expected markup', () => {
  const wrapper = mount(<Text />);
  expect(wrapper).toMatchSnapshot();
});

it('renders a component with a different tag', () => {
  const wrapper = mount(<Text as="h1" type="xlarge" />);
  expect(wrapper).toMatchSnapshot();
});

it('renders the plain skeleton', () => {
  const wrapper = mount(<Text.Skeleton>foobar foobaz</Text.Skeleton>);
  expect(wrapper).toMatchSnapshot();
});

it('renders a Skeleton with a different tag', () => {
  const wrapper = mount(<Text.Skeleton as="h1">foobar foo</Text.Skeleton>);
  expect(wrapper).toMatchSnapshot();
});

it('renders a Skeleton with set width', () => {
  const wrapper = mount(<Text.Skeleton width="20px" />);
  expect(wrapper).toMatchSnapshot();
});
