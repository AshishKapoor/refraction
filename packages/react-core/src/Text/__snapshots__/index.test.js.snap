// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a Skeleton with a different tag 1`] = `
@keyframes animation-0 {
  0% {
    background-color: #C7CCD1;
  }

  50% {
    background-color: rgba(199,204,209,0.5);
  }

  100% {
    background-color: #C7CCD1;
  }
}

.emotion-0 {
  color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 2px;
  background-color: #C7CCD1;
  -webkit-animation: animation-0 3s ease-in-out infinite;
  animation: animation-0 3s ease-in-out infinite;
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
}

<Skeleton
  as="h1"
>
  <h1>
    <span
      className="emotion-0 emotion-1"
    >
      foobar foo
    </span>
  </h1>
</Skeleton>
`;

exports[`renders a Skeleton with set width 1`] = `
@keyframes animation-0 {
  0% {
    background-color: #C7CCD1;
  }

  50% {
    background-color: rgba(199,204,209,0.5);
  }

  100% {
    background-color: #C7CCD1;
  }
}

@keyframes animation-0 {
  0% {
    background-color: #C7CCD1;
  }

  50% {
    background-color: rgba(199,204,209,0.5);
  }

  100% {
    background-color: #C7CCD1;
  }
}

.emotion-0 {
  color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 2px;
  background-color: #C7CCD1;
  -webkit-animation: animation-0 3s ease-in-out infinite;
  animation: animation-0 3s ease-in-out infinite;
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  width: 20px;
  display: inline-block;
}

.emotion-0::before {
  content: 'X';
}

<Skeleton
  width="20px"
>
  <span>
    <span
      className="emotion-0 emotion-1"
    />
  </span>
</Skeleton>
`;

exports[`renders a component with a different tag 1`] = `
.emotion-0 {
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
  font-size: 20px;
  line-height: 1.4;
}

<Text
  as="h1"
  type="xlarge"
>
  <h1
    className="emotion-0 emotion-1"
    type="xlarge"
  />
</Text>
`;

exports[`renders the expected markup 1`] = `
.emotion-0 {
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
}

<Text>
  <span
    className="emotion-0 emotion-1"
  />
</Text>
`;

exports[`renders the plain skeleton 1`] = `
@keyframes animation-0 {
  0% {
    background-color: #C7CCD1;
  }

  50% {
    background-color: rgba(199,204,209,0.5);
  }

  100% {
    background-color: #C7CCD1;
  }
}

.emotion-0 {
  color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border-radius: 2px;
  background-color: #C7CCD1;
  -webkit-animation: animation-0 3s ease-in-out infinite;
  animation: animation-0 3s ease-in-out infinite;
  font-family: IBM Plex Sans,Lucida Grande,Tahoma,Verdana,Arial,sans-serif;
}

<Skeleton>
  <span>
    <span
      className="emotion-0 emotion-1"
    >
      foobar foobaz
    </span>
  </span>
</Skeleton>
`;
